example.env

```diff
@@ -192,6 +192,21 @@ PYCHARM_REMOTE_DEBUG_SERVER_PORT_GUI=55875
 ## Path string for Pycharm remote debug package egg
 #PYCHARM_REMOTE_DEBUG_EGG_PATH=
 
+################################################################################
+################################################################################
+## Numpy Settings                                                             ##
+################################################################################
+################################################################################
+
+## Version of Numpy to select when installing (e.g., within Docker images)
+## Note that this probably needs to stay in sync with what is in a few other places
+##  that don't read directly from this file:
+##      - requirement.txt
+##      - python/gui/dependencies.txt
+# Note: mind the consecutive equal signs; the value being set here is '==1.26.4'
+NUMPY_INSTALL_CONSTRAINT===1.26.4
+
+
```

Comment:
This is an abuse of the args parameter. Do not use this to inject python versions. Stick to setup.py and requirements.txt.

The use of the `.env` pattern here needs to be reduced, not added to.

Reply:
> The use of the .env pattern here needs to be reduced, not added to.

I don't think I agree with this exactly as intended.  Our structures for configuring a DMOD environment need to be improved; using .env files like this is not the most effective thing we could be doing.  But that shouldn't equate to reducing what users are able to configure, or restricting ourselves from adding more configurable options when they are needed.

But, setting aside how we implement things, should we simply not allow users to control what version of numpy is installed in worker images?  If not, why?

Reply:
If users desire to change versions of library dependencies they are coloring that far outside the lines they need to do so via channels like their requirements.txt, setup.py files, or their own versions of their build scripts and files.  This degree of configuration is too much and provides too many avenues for failure.

Reply:
> If users desire to change versions of library dependencies they are coloring that far outside the lines they need to do so via channels like their requirements.txt, setup.py files, or their own versions of their build scripts and files.

I worry there will be a particular problem here when we start to bring in BMI modules that aren't directly integrated into ngen or the DMOD image (which is coming soon).  For a users to essentially dynamically inject a BMI module for use within job workers, it's very likely that at least some of them will need the ability to control the version of numpy installed (i.e., when ngen is compiled).  

We could, alternatively, disallow (formally or de facto) any injected modules if they don't support a numpy version hard-coded into the worker.  In principle, I think that hurts community adoption; in practice, maybe the scenario doesn't happen frequently enough to outweigh other concerns.

Reply:
As it stands, the requirements of "env file + environment variables + multiple Dockerfiles + Swarm" will hurt adoption on its own. 

What are the plans for dynamically building images with specific numpy versions if the the lever for the setting is global for the entire runtime environment? Is it going to require direct calls to `docker build` with a long list of `arg` parameters? If so, does that also indicate that the process of building images has _also_ become too cumbersome?

When these requirements appear, a solution for that needs to be implemented using different strategies, such as adding `COPY` hook into the Dockerfile to allow for more advanced behavior for more advanced BMI modules. That, in of itself, is easier than juggling `ARGS` within the image definitions.

Reply:
> As it stands, the requirements of "env file + environment variables + multiple Dockerfiles + Swarm" will hurt adoption on its own.

I'd be shocked if one more configurable .env setting hurt adoption any more than using that approach for deployment configuration to begin with.  And I'm not suggesting we don't overhaul how deployments are configured, but that's way out of scope for this.

> What are the plans for dynamically building images with specific numpy versions if the the lever for the setting is global for the entire runtime environment?

I'm not really sure what you are asking here.  If you're talking about ways to customize different tags of the worker image with potentially different versions of numpy, I wasn't going to get into that here.  This was just to fix something that was broken (with the solution being a new knob to turn, rather than something hard-coded, since I expected we might need that knob soon anyway for other reasons).  But I'm sure once we start to get into #661 we'd need to think about things like this.